src/ports/ORMPUpgradeableAndRetryablePort.sol

```diff
@@ -0,0 +1,82 @@
+// This file is part of Darwinia.
+// Copyright (C) 2018-2023 Darwinia Network
+// SPDX-License-Identifier: GPL-3.0
+//
+// Darwinia is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Darwinia is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.17;
+
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "ORMP/src/user/UpgradeableApplication.sol";
+import "./ORMPPort.sol";
+
+contract ORMPUpgradeableAndRetryablePort is ORMPPort, UpgradeableApplication, ReentrancyGuard {
+    /// @dev msgHash => isDispathedInPort.
+    mapping(bytes32 => bool) public dones;
+
+    event MessageDispatchedInPort(bytes32 indexed msgHash);
+    event ClearFailedMessage(bytes32 indexed msgHash);
+
+    constructor(address dao, address ormp, string memory name) ORMPPort(dao, ormp, name) UpgradeableApplication(ormp) {}
+
+    function protocol() public view override(Application, UpgradeableApplication) returns (address) {
+        return super.protocol();
+    }
+
+    function setORMP(address ormp) external onlyOwner {
+        _setORMP(ormp);
+    }
+
+    function retryFailedMessage(Message calldata message) external payable nonReentrant {
```

Comment:
DOC: retry in Dapp layer.

Reply:
1. gas limit estimate issue cause Dapp failed.
2. dapp buz failed need retry or clear message.

Reply:
1. Move retryable feature to RetryableApp template.
    - ORMP.dones(msgHash)
    - ormpport_ext.checkDeliveryed()
3. Keep upgradeable feature.

Reply:
1. provide retryable port and non-retrayable port for dapp chosen.
2. If dapp choose retryable port, dapp need to implement `IMessageRetrayablePort.sol` interface.